hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/common/ChunkBufferImplWithByteBufferList.java

```diff
@@ -0,0 +1,215 @@
+  private void findCurrent() {
+    boolean found = false;
+    for (int i = 0; i < buffers.size(); i++) {
+      final ByteBuffer buf = buffers.get(i);
+      final int pos = buf.position();
+      if (found) {
+        Preconditions.checkArgument(pos == 0,
+            "all buffers after current one should have position=0");
+      } else if (pos < buf.limit()) {
+        found = true;
+        currentIndex = i;
+      } else {
+        limitPrecedingCurrent += buf.limit();
+      }
+    }
+    if (!found) {
+      currentIndex = buffers.size() - 1;
+      limitPrecedingCurrent -= current().limit();
+    }
+  }
+
+  private ByteBuffer current() {
+    return buffers.get(currentIndex);
+  }
+
+  private void advanceCurrent() {
+    if (currentIndex < buffers.size() - 1) {
+      final ByteBuffer current = buffers.get(currentIndex);
+      if (!current.hasRemaining()) {
+        currentIndex++;
+        limitPrecedingCurrent += current.limit();
+      }
+    }
+  }
+
+  private void rewindCurrent() {
+    currentIndex = 0;
+    limitPrecedingCurrent = 0;
+  }
+
+  @Override
+  public int position() {
+    return limitPrecedingCurrent + current().position();
+  }
+
+  @Override
+  public int remaining() {
+    return limit - position();
+  }
+
+  @Override
+  public int limit() {
+    return limit;
+  }
+
+  @Override
+  public ChunkBuffer rewind() {
+    buffers.forEach(ByteBuffer::rewind);
+    rewindCurrent();
+    return this;
+  }
+
+  @Override
+  public ChunkBuffer clear() {
+    buffers.forEach(ByteBuffer::clear);
+    rewindCurrent();
+    return this;
+  }
+
+  @Override
+  public ChunkBuffer put(ByteBuffer that) {
+    final int thisRemaining = remaining();
+    int thatRemaining = that.remaining();
+    if (thatRemaining > thisRemaining) {
+      final BufferOverflowException boe = new BufferOverflowException();
+      boe.initCause(new IllegalArgumentException(
+          "Failed to put since that.remaining() = " + thatRemaining
+              + " > this.remaining() = " + thisRemaining));
+      throw boe;
+    }
+
+    while (thatRemaining > 0) {
+      final ByteBuffer b = current();
+      final int bytes = Math.min(b.remaining(), thatRemaining);
+      that.limit(that.position() + bytes);
+      b.put(that);
+      thatRemaining -= bytes;
+      advanceCurrent();
+    }
+
+    return this;
+  }
+
+  @Override
+  public ChunkBuffer duplicate(int newPosition, int newLimit) {
+    Preconditions.checkArgument(newPosition >= 0);
+    Preconditions.checkArgument(newPosition <= newLimit);
+    Preconditions.checkArgument(newLimit <= limit());
+
+    final List<ByteBuffer> duplicates = new ArrayList<>(buffers.size());
+    int min = 0;
+    for (final ByteBuffer buf : buffers) {
+      final int max = min + buf.limit();
+      final int pos = relativeToRange(newPosition, min, max);
+      final int lim = relativeToRange(newLimit, min, max);
+
+      final ByteBuffer duplicate = buf.duplicate();
+      duplicate.position(pos).limit(lim);
```

Comment:
I think position and limit are set with buf.duplicate(). We might not need the computation of min, max and lim in that case.

Reply:
Calculation of min/max/lim is required because `ChunkBuffer#duplicate` combines operation of `duplicate` and `position`/`limit` (by accepting `newLimit` and `newPosition` parameters).